lib/MongoDB/_Protocol.pm

```diff
@@ -139,48 +138,155 @@ sub prepare_sections {
     delete => 'deletes',
   );
 
-  my ( $command, $collection, $ident, $docs, @other ) = @$cmd;
+  $cmd = to_IxHash( $cmd );
+
+  # Command is always first key in cmd
+  my $command = do { my @keys = $cmd->Keys; $keys[0] };
+  my $ident = $split_commands{ $command };
 
-  if ( $split_commands{ $command } eq $ident ) {
-    # Assumes only a single split on the commands
+  if ( defined $ident
+    && $split_commands{ $command } eq $ident ) {
+    my $collection = $cmd->FETCH( $command );
+    my $docs = $cmd->FETCH( $ident );
+    # my ( undef, $collection ) = $cmd->Shift;
+    # my ( undef, $docs ) = $cmd->Shift;
+    # # Assumes only a single split on the commands
     return (
-      MongoDB::Protocol::_Section->new(
-        bson_codec => $codec,
-        type => 0,
-        documents => [ [ $command, $collection, @other ] ]
-      ),
-      MongoDB::Protocol::_Section->new(
-        bson_codec => $codec,
-        type => 1,
-        identifier => $ident,
-        documents => $docs
-      ),
+        {
+            type => 0,
+            documents => [ [
```

Comment:
I actually named it documents to keep it with the specification - until they're actually packed they are not a payload object, but the documents which are the content of the payload. if it was going to be more accurate to the spec it would be something like:

```
{
  type => $type,
  payload => {
    (defined $ident ? (identifier => $ident) : () ),
    documents => $docs
  }
}
```

but payload there seems a bit of a redundant key...